packages/commonwealth/server/controllers/server_profiles_methods/get_community_members.ts

```diff
@@ -0,0 +1,304 @@
+  this: ServerProfilesController,
+  {
+    communityId,
+    search,
+    includeRoles,
+    limit,
+    page,
+    orderBy,
+    orderDirection,
+    memberships,
+    includeGroupIds,
+    includeStakeBalances,
+  }: GetMemberProfilesOptions,
+): Promise<GetCommunityMembersResult> {
+  const community = await this.models.Community.findByPk(communityId);
+  if (!community) {
+    throw new AppError(Errors.CommunityNotFound);
+  }
+  page = Math.min(1, page);
+
+  const bind: any = {
+    searchTerm: `%${search}%`,
+  };
+  if (community) {
+    bind.community_id = community.id;
+  }
+
+  const communityWhere = bind.community_id
+    ? `"Addresses".community_id = $community_id AND`
+    : '';
+
+  const groupIdFromMemberships = parseInt(
+    ((memberships || '').match(/in-group:(\d+)/) || [`0`, `0`])[1],
+  );
+  let membershipsWhere = memberships
+    ? `SELECT 1 FROM "Memberships"
+    JOIN "Groups" ON "Groups".id = "Memberships".group_id
+    WHERE "Memberships".address_id = "Addresses".id
+    AND "Groups".community_id = $community_id
+    ${
+      groupIdFromMemberships
+        ? `AND "Groups".id = ${groupIdFromMemberships}`
+        : ''
+    }
+    `
+    : '';
+
+  if (memberships) {
+    switch (memberships) {
+      case 'in-group':
+      case `in-group:${groupIdFromMemberships}`:
+        membershipsWhere = `AND EXISTS (${membershipsWhere} AND "Memberships".reject_reason IS NULL)`;
+        break;
+      case 'not-in-group':
+        membershipsWhere = `AND NOT EXISTS (${membershipsWhere} AND "Memberships".reject_reason IS NULL)`;
+        break;
+      default:
+        throw new AppError(`unsupported memberships param: ${memberships}`);
+    }
+  }
+
+  const sqlWithoutPagination = `
+    SELECT
+      "Profiles".id,
+      "Profiles".user_id,
+      "Profiles".profile_name,
+      "Profiles".avatar_url,
+      "Profiles".created_at,
+      array_agg("Addresses".id) as address_ids,
+      array_agg("Addresses".community_id) as chains,
+      array_agg("Addresses".address) as addresses,
+      MAX("Addresses".last_active) as last_active
+    FROM
+      "Profiles"
+    JOIN
+      "Addresses" ON "Profiles".user_id = "Addresses".user_id
+    WHERE
+      ${communityWhere}
+      (
+        "Profiles".profile_name ILIKE '%' || $searchTerm || '%'
+        OR
+        "Addresses".address ILIKE '%' || $searchTerm || '%'
+      )
+      ${membershipsWhere}
+    GROUP BY
+      "Profiles".id
+  `;
+
+  const fullResults = await this.models.sequelize.query<{
+    id: number;
+    user_id: number;
+    profile_name: string;
+    avatar_url: string;
+    created_at: string;
+    address_id: string[];
+    chains: string[];
+    addresses: string[];
+    stake_balances: string[];
+    last_active: string;
+  }>(`${sqlWithoutPagination}`, {
+    bind,
+    type: QueryTypes.SELECT,
+  });
+  const totalResults = fullResults.length;
+
+  if (includeStakeBalances) {
+    const stake = await this.models.CommunityStake.findOne({
```

Comment:
NIT: a minor optimization, we can consolidate the 3 queries below like:

```
SELECT g.*, cn.id FROM "CommunityStakes" AS cs
JOIN "Groups" AS g ON g.community_id = cs.community_id
RIGHT JOIN "ChainNodes" as cn ON cn.community_id = cs.community_id
WHERE g.is_system_managed = true and g.community_id = ${community_id}
```

Same argument can be made (Excluding Groups join) for generate_stakeholder_group CommunityStake/ChainNode check.

Reply:
These are all great suggestions. Will apply in a FF PR since this needs to get in.